src/main/groovy/org/jenkinsci/gradle/plugins/jpi/JpiPlugin.groovy

```diff
@@ -349,4 +346,38 @@ class JpiPlugin implements Plugin<Project> {
         }
         project.tasks.test.dependsOn(generateTestHpl)
     }
+
+    private static void resolvePluginDependencies(Project project) {
+        resolvePluginDependencies(
+                project,
+                PLUGINS_DEPENDENCY_CONFIGURATION_NAME,
+                PROVIDED_COMPILE_CONFIGURATION_NAME
+        )
+        resolvePluginDependencies(
+                project,
+                OPTIONAL_PLUGINS_DEPENDENCY_CONFIGURATION_NAME,
+                PROVIDED_COMPILE_CONFIGURATION_NAME
+        )
+        resolvePluginDependencies(
+                project,
+                JENKINS_TEST_DEPENDENCY_CONFIGURATION_NAME,
+                TEST_COMPILE_CONFIGURATION_NAME
+        )
+    }
+
+    private static void resolvePluginDependencies(Project project, String from, String to) {
+        ConfigurationContainer configurations = project.configurations
+        Configuration fromConfiguration = configurations.getByName(from)
+        Configuration toConfiguration = configurations.getByName(to)
+
+        project.configurations.each { Configuration c ->
+            c.incoming.beforeResolve { ResolvableDependencies resolvableDependencies ->
+                if (c.hierarchy.contains(toConfiguration) && fromConfiguration.state == UNRESOLVED) {
+                    fromConfiguration.resolvedConfiguration.resolvedArtifacts
+                            .findAll { it.type == 'hpi' || it.type == 'jpi' }
+                            .each { project.dependencies.add(to, "${it.moduleVersion.id}@jar") }
```

Comment:
IIUC, we then have the `jpi` dependency and the `jar` dependency on the `to` configuration, given that these configurations extend from the `from` configurations. Is that what we really want? Shouldn't we better have only the `jar`s in there?

Reply:
Good point. When I remote the calls to `extendsFrom`, job-dsl-plugin has test failures due to `java.lang.ClassNotFoundException`. Not sure why, I need to compare the dependencies. But does it hurt to have `jpi` and `jar` dependencies in the same configuration?

Reply:
I guess it does not hurt. I think Java 9 has problems with files on the classpath which are not zips, but that is also not problematic here. I think the `ClassNotFoundExceptions` come from the fact that we only add the `jpi`/`hpi` artifacts here, but not the regular `jar` artifacts in the configurations. So I guess it should be something like:
```
fromConfiguration.resolvedConfiguration.resolvedArtifacts
    .each {
        if (it.type == 'hpi' || it.type == 'jpi') {
            project.dependencies.add(to, "${it.moduleVersion.id}@jar")
        } else {
            project.dependencies.add(to, it)
        }
    }
```

Reply:
Ah, you are right. That was missing.

As it does not seem to be problematic, I would merge this PR as it is and deal with separating `jar` and `jpi` configurations in another PR. The `jenkinsTest` configuration currently contains both, the plugins for testing (`jpi`/`hpi' ) and the Jenkins Test Harness (`jar`), so we would need to split that to get a clean separation.

I'll file a JIRA issue for separating the configuration, merge this PR and cut a release if there are no objections.

Reply:
👍 SGTM!

Reply:
The JIRA issue for separating the configurations: https://issues.jenkins-ci.org/browse/JENKINS-47375